google/cloud/internal/win32/parse_service_account_p12_file.cc

```diff
@@ -0,0 +1,318 @@
+        GCP_ERROR_INFO());
+  }
+  return UniqueCertStore(certstore_raw);
+}
+
+StatusOr<UniqueCertContext> GetCertificate(HCERTSTORE certstore,
+                                           std::string const& source) {
+  // Get the certificate from the store.
+  PCCERT_CONTEXT cert_raw = CertFindCertificateInStore(
+      certstore, X509_ASN_ENCODING | PKCS_7_ASN_ENCODING, 0, CERT_FIND_ANY,
+      nullptr, nullptr);
+  if (cert_raw == nullptr) {
+    return InvalidArgumentError(
+        absl::StrCat("No certificate found in PKCS#12 file (", source, ")"),
+        GCP_ERROR_INFO());
+  }
+  return UniqueCertContext(cert_raw);
+}
+
+std::string GetCertificateCommonName(PCCERT_CONTEXT cert) {
+  DWORD size = CertGetNameStringA(cert, CERT_NAME_SIMPLE_DISPLAY_TYPE, 0,
+                                  nullptr, nullptr, 0);
+  std::string result(size, '\0');
+  CertGetNameStringA(cert, CERT_NAME_SIMPLE_DISPLAY_TYPE, 0, nullptr,
+                     &result[0], size);
+  result.pop_back();  // remove the null terminator
+  return result;
+}
+
+StatusOr<UniqueCryptProv> GetCertificatePrivateKey(PCCERT_CONTEXT cert,
+                                                   DWORD& dwKeySpec,
+                                                   std::string const& source) {
+  HCRYPTPROV prov_raw;
+  BOOL pfCallerFreeProvOrNCryptKey;
+  if (!CryptAcquireCertificatePrivateKey(cert, CRYPT_ACQUIRE_SILENT_FLAG,
+                                         nullptr, &prov_raw, &dwKeySpec,
+                                         &pfCallerFreeProvOrNCryptKey)) {
+    return InvalidArgumentError(
+        FormatWin32Errors("No private key found in PKCS#12 file (", source,
+                          "): "),
+        GCP_ERROR_INFO());
+  }
+  // According to documentation of CryptAcquireCertificatePrivateKey,
+  // pfCallerFreeProvOrNCryptKey will always be true in our case so
+  // we don't need to check it.
+  return UniqueCryptProv(prov_raw);
+}
+
+StatusOr<UniqueCryptKey> GetKeyFromProvider(HCRYPTPROV prov, DWORD dwKeySpec,
+                                            std::string const& source) {
+  HCRYPTKEY pkey_raw;
+  if (!CryptGetUserKey(prov, dwKeySpec, &pkey_raw)) {
+    return InvalidArgumentError(
+        FormatWin32Errors("No private key found in PKCS#12 file (", source,
+                          "): "),
+        GCP_ERROR_INFO());
+  }
+  return UniqueCryptKey(pkey_raw);
+}
+
+StatusOr<std::vector<char>> ExportPrivateKey(HCRYPTKEY pkey,
+                                             std::string const& source) {
+  DWORD exported_key_length;
+  if (!CryptExportKey(pkey, 0, PRIVATEKEYBLOB, 0, nullptr,
+                      &exported_key_length)) {
+    return InvalidArgumentError(
+        FormatWin32Errors("Could not export private key from PKCS#12 file (",
+                          source, "): "),
+        GCP_ERROR_INFO());
+  }
+  std::vector<char> exported_key(exported_key_length);
```

Comment:
Ditto, should this be `std::vector<BYTE>`?

Reply:
Changed.